### PR TITLE
Adaptive queue for staging dials

### PR DESCRIPTION
**Pull request by raulk**
Currently the DHT is performing dials outside of the Alpha concurrency limit. We are dialling all nodes that peers return in `CloserPeers` without limit. As a result, we end up flooding the swarm with dial jobs, which trips over the file descriptor limits, and brings dialling to a halt under some circumstances. Our current approach is also algorithmically incorrect, and leads to suboptimal query patterns.

This patch introduces an adaptive dial queue that spawns a dynamically sized set of goroutines to preemptively stage dials for later handoff to the DHT protocol for RPC. It identifies backpressure on both ends (dial consumers and dial producers), and takes compensating action by adjusting the worker pool.

We start with `DialQueueMinParallelism` number of workers (6), and scale up and down based on demand and supply of dialled peers.

The following events trigger scaling:
- we scale up when we can't immediately return a successful dial to a new consumer.
- we scale down when we've been idle for a while waiting for new dial attempts.
- we scale down when we complete a dial and realise nobody was waiting for it.

Dialler throttling (e.g. FD limit exceeded) is a concern, as we can easily spin up more workers to compensate, and end up adding fuel to the fire. Since we have no deterministic way to detect this for now, we hard-limit concurrency to `DialQueueMaxParallelism` (20).

### CODE REVIEW

File: query.go [hunk 1]
```diff
@@ -70,10 +74,207 @@ func (q *dhtQuery) Run(ctx context.Context, peers []peer.ID) (*dhtQueryResult, e
...
+	sq := &stagingQueue{
+		dialFn:        dialFn,
+		ctx:           ctx,
+		scalingFactor: 1.5,
+		parallelism:   StagingQueueMinParallelism,
+		in:            in,
+
+		// TODO: we should really re-sort dials as they come back based on XORKeySpace, instead of queuing them in the
+		// order they come back.
+		completedCh: make(chan peer.ID),
```
Comment (Stebalien) [inline query.go]: We could use another priority queue. However, we'll need some other backpressure mechanism, some kind of ticket system?

**Comment by Stebalien** [comment]
I'm worried the backpressure mechanism may be backwards at the extreme. That is, up to a point, increasing the dial parallelism will increase the rate at which we can query peers. However, when the dial limiters kick in (possibly due to other parallel processes outside of our control), adding more dials will make things worse (so we'll keep adding more dials).

**Comment by raulk** [comment]
TODO:

- [x] Resorting successful dials based on XOR distance function, instead of delivering them in the order they finished.
- [x] ~~Using dial waiting time moving average as a source of backpressure. When we have many dials inflight and they've been parked for a while, we're probably being rate-limited by the dialer, and scaling up is counterproductive. If dials are progressing but only a few are stalled, they are probably slow peers, and it's fair to scale up.~~ We cannot detect FD throttling deterministically.
- [x] Try to avoid the goroutine per each consumption. Maybe find a way for workers to publish directly to consumers, instead of going through completedCh.
- [x] Debounce scaling signals to avoid thrashing.

File: query.go [hunk 2]
```diff
@@ -70,10 +74,211 @@ func (q *dhtQuery) Run(ctx context.Context, peers []peer.ID) (*dhtQueryResult, e
...
+
+	in  *queue.ChanQueue
+	out *queue.ChanQueue
+
+	waiting chan chan<- peer.ID
+	dieCh   chan struct{}
+}
+
+var DialQueueMinParallelism = 6
+var DialQueueMaxParallelism = 20
```
Comment (Stebalien) [inline query.go]: We should check to see if we floor/ceil against one of these. If so, that would indicate that the extra complexity may not be worth it.

**Review by Stebalien — COMMENTED** [review]
I haven't checked this thoroughly for bugs but it *looks* right. It's also pretty easy to follow. I was worried we'd have to introduce a large maintenance burden to fix this...

File: query.go (hunk 2 above)
Comment (Stebalien) [reply query.go]: After reading this, it doesn't look too complex anyways. However, we should still check this.

**Comment by raulk** [comment]
@Stebalien would appreciate a pass on the latest changes. Next step is to test them out in a local instance of IPFS.

I'm also wondering if introducing a ticketing system would give us a better handle on the scaling. Right now it's kinda clunky and event-based:

* we scale up when we can't immediately return a successful dial to a new consumer.
* we scale down when we've been idle for a while waiting for new dial attempts.
* we scale down when we complete a dial and realise nobody was waiting for it.

To avoid excessive jitter, there's a mute period of 1 second between scaling events.

With a ticketing system we could make it time-based and more sophisticated. We could monitor:
* how long consumers are waiting before an item is delivered to them.
* how long successful dials are parked before a consumer asks for them.

If consumers are waiting too long (WIP: define what is too long), we spin more workers. If dials are parked for too long, we kill workers. Thus balancing would be in flux (e.g. every second) by identifying positive and negative backpressure and compensating it.

File: dial_queue.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+import (
+	"context"
+	"errors"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-peer"
+	"github.com/libp2p/go-libp2p-peerstore/queue"
```
Comment (ghost) [inline dial_queue.go]: Could you name the imports? Otherwise it's not obvious which identifier they end up with.

Comment (raulk) [reply dial_queue.go]: Sure, this is not the official PR, just a venue to discuss the solution, but good call!

**Comment by anacrolix** [comment]
It's very elaborate. It might just be easier to be allocated some headroom from the connection manager, and dial on demand within that quota with timeouts.

**Comment by raulk** [comment]
@anacrolix Unfortunately, the connection manager does not intervene here; the limitation is on inflight dials (not running connections), i.e. swarm.

We did consider a simpler solution: separate buffers for dials and RPC, with dials being 5xAlpha. Unfortunately that does not behave well in realistic scenarios, like picking 15 slow/undialable peers, or peers with address pollution (we have peers with >100 addresses in the DHT).

File: dial_queue.go (hunk 3 above)
Comment (raulk) [reply dial_queue.go]: ✅

**Review by anacrolix — APPROVED** [review]

**Comment by raulk** [comment]
Looks like we are reaching consensus on this, so I'm going to transplant this PR to the main repo.

**Pull request closed without merging**